### PR TITLE
feat(favouriteStations): Enable users select their favourite stations and filter them

### PR DESCRIPTION
**Pull request by dimitrantz**
Implement a new feature that allows the users to store their favourite stations cached in the browser. This solution leverages the localstorage feature since there is no concept of a single user just yet.

Additional notes:
The user can add a station to their feavourite list via the station modal. A new heart icon has been added to support this action as shown below:
image

In case there is at least one station added in the list of favourite stations, then a new toggle becomes visible so the user can filter the stations in the Weather Map and Stations pages. Please see the images below:

WeatherMap page:
_When the toggle is disabled:_
image
_When the toggle is enabled:_
image

_The search form is being updated dynamically without the need of a page reload to reflect the filter:_
image

Stations page:
_When the toggle is disabled:_
image
_When the toggle is enabled:_
image

The feature works for multiple browsers windows and tabs, however I've kept the solution minimal for now and have not adde any listeners. This means if the user has opened multiple windows/tabs, any changes in one of them will require a page reload in the other to reflect the changes. This could be improved further if required in a future task.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -26,6 +26,7 @@
     "dependencies": {
         "@headlessui/react": "1.7.17",
         "@heroicons/react": "2.0.18",
+        "@material-tailwind/react": "2.1.10",
```
Comment (Virtuosofriend) [inline package.json:29]: I would suggest to use https://www.radix-ui.com/ instead so we don't use multiple component libraries (I have already used the slider primitive one from them).
I was thinking about switching over to them for every UI component so we don't have to look for custom solutions all the time. What do you think?

File: src/components/Common/ToggleSwitch.tsx [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+    const [isMounted, setIsMounted] = useState(false);
+
+    useEffect(() => {
+        setIsMounted(true);
+    }, []);
+
+    if (!isMounted) {
+        return null; // Ensure consistent server-client output
+    }
```
Comment (Virtuosofriend) [inline src/components/Common/ToggleSwitch.tsx:20]: I don't get the why here. I believe it has something to do with React/next rendering but I am afraid I don't understand it.

File: src/components/Header/HeaderMenu.tsx [hunk 3]
```diff
@@ -1,21 +1,31 @@
...
+                        }`}
+                        href={element.pathName}
+                        key={element.text}
+                    >
+                        {element.text}
+                    </Link>
+                ))}
+            </div>
+
+            {favouriteStations.length > 0 && <ToggleSwitch label={"My Favourite Stations"} />}
```
Comment (Virtuosofriend) [inline src/components/Header/HeaderMenu.tsx:28]: This addition here is my main objection to this project. I do feel it is too much to have this in the header as an option. 
The header is common for all the pages at the moment as a result the switch will be also visible in the warnings page (and in the future about us page or any other irrelevant one). Maybe it would be better to move it to the necessary views only?

Comment (Virtuosofriend) [reply src/components/Header/HeaderMenu.tsx:28]: On the sage topic, I was thinking about adding options like windy does on the right of the map screen. That space is occupied by warnings at the moment (desktop) which should be revamped, maybe we could use the favorite station feature as the breaking ground for this. Or maybe, we could merge this (not push it to prod just yet) while I revamp the warnings feature.

File: src/components/Header/MobileHeaderMenu.tsx [hunk 4]
```diff
@@ -1,36 +1,39 @@
+import { useAppStore } from "@/hooks/useAppStore";
```
Comment (Virtuosofriend) [inline src/components/Header/MobileHeaderMenu.tsx:1]: Same comments regarding the menu as above.

File: src/components/Home/HomepageMap.tsx [hunk 5]
```diff
@@ -67,37 +67,47 @@ const createClusterCustomIcon = function (cluster: MarkerCluster) {
     });
 };
 
+const getStationsMarkers = function (
```
Comment (Virtuosofriend) [inline src/components/Home/HomepageMap.tsx:70]: Out of curiosity, why did you move it outside the `HomepageMap`?

File: src/components/Stations/StationsPage.tsx [hunk 6]
```diff
@@ -1,16 +1,26 @@
-import { useState, useEffect } from "react";
+import { useAppStore } from "@/hooks/useAppStore";
```
Comment (Virtuosofriend) [inline src/components/Stations/StationsPage.tsx:1]: General proposal for this page.
My original idea regarding this page was to give the user a general outlook of all the stations. Maybe, we could add a fav stations section on top of the table, with all the fav stations there and the rest below. Just shooting ideas here.

File: package.json (hunk 1 above)
Comment (dimitrantz) [reply package.json:29]: You were right, I didn't use this module at the end 😶‍🌫️ 
Haven't tried [them](https://www.radix-ui.com/) but had a look and they seem cool, let's have a ticket to switch all of them in all UI components. Can pick this up after the fav feature.

File: src/components/Home/HomepageMap.tsx (hunk 5 above)
Comment (dimitrantz) [reply src/components/Home/HomepageMap.tsx:70]: As discussed offline, just moved it outside to improve performace since we don't need to define it every time we render the component

File: src/components/Common/ToggleSwitch.tsx (hunk 2 above)
Comment (dimitrantz) [reply src/components/Common/ToggleSwitch.tsx:20]: As discussed this was added to resolve compilation issues due to using the localstorage in the client side before it was properly initialised on the server side. We're completely removing the toggle in the new approach and as as discussed might be better to use the persist store so I'll explore that instead.

File: src/components/Header/MobileHeaderMenu.tsx (hunk 4 above)
Comment (dimitrantz) [reply src/components/Header/MobileHeaderMenu.tsx:1]: Yep, the toggle is completely removed from the header now

**Comment by dimitrantz** [comment]
Hiya, please find below some screenshots after applying the suggestions:

When the user has no favourite stations:
image
image

When the user has favourite stations:
image
image


image

File: src/components/BaseComponents/BaseModal.tsx [hunk 7]
```diff
@@ -37,12 +41,24 @@ export default function BaseModal({
...
                             leaveTo="opacity-0 scale-95"
                         >
-                            <Dialog.Panel className={`min-h-[430px] w-full overflow-hidden rounded-2xl bg-white p-2 text-left shadow-xl transition-all ${dialogClass}`}>
+                            <Dialog.Panel
+                                className={`min-h-[430px] w-full overflow-hidden rounded-2xl bg-white p-2 text-left shadow-xl transition-all ${dialogClass}`}
+                            >
                                 <Dialog.Description as="section">
                                     {children}
-                                    <div className="absolute right-3 top-3">
+                                    <div className="absolute right-3 top-3 flex gap-2">
```
Comment (Virtuosofriend) [inline src/components/BaseComponents/BaseModal.tsx:49]:
```suggestion
                                    <div className="absolute right-3 top-3 flex gap-2 items-center">
```
Probably the two icons being identical in height won't create any misalignment but better be safe than sorry.

File: src/components/Header/HeaderMenu.tsx [hunk 8]
```diff
@@ -1,21 +1,26 @@
...
-                key={element.text}
-            >
-                {element.text}
-            </Link>
-        )}
-    </section>
+
+    return (
+        <section className="flex flex-col">
+            <div className="mt-6 flex items-center gap-4 lg:gap-0">
```
Comment (Virtuosofriend) [inline src/components/Header/HeaderMenu.tsx:11]: Probably leftover after the removal of the switch?

Comment (dimitrantz) [reply src/components/Header/HeaderMenu.tsx:11]: yep, removed

File: src/components/BaseComponents/BaseModal.tsx (hunk 7 above)
Comment (dimitrantz) [reply src/components/BaseComponents/BaseModal.tsx]: sure, updated

File: src/components/Stations/components/StationTableColumns.tsx [hunk 9]
```diff
@@ -0,0 +1,165 @@
...
+                            handleClick={() => handleFavouriteButton(info.getValue())}
+                            color={isStationFavourite(info.getValue()) ? "primary" : "secondary"}
+                        >
+                            <HeartIcon className="size-8 p-1" />
+                        </CommonButton>
+                    </div>
+                );
+            },
+            header: () => <span>Favourite</span>,
+            sortingFn: sortStatusFn,
```
Comment (Virtuosofriend) [inline src/components/Stations/components/StationTableColumns.tsx:42]: I would disable sorting to be honest for this column.

File: src/components/Stations/components/StationTableColumns.tsx [hunk 10]
```diff
@@ -0,0 +1,165 @@
...
+
+// Columns initialisation
+const columnHelper = createColumnHelper<WeatherDataResponse>();
+export const getColumns = (
+    title: string,
+    handleFavouriteButton: (stationId: number) => void,
+    isStationFavourite: (stationId: number) => boolean
+) => {
+    return [
+        columnHelper.accessor((row) => row.weather_station_id.id, {
```
Comment (Virtuosofriend) [inline src/components/Stations/components/StationTableColumns.tsx:27]: I would also move this column to the end of the table. In mobile view, the user seems to lose quite valuable real estate for this single column. On top of that, usually actions are located at the end of rows in data tables.

File: src/components/Stations/components/StationTableColumns.tsx [hunk 11]
```diff
@@ -0,0 +1,165 @@
+import BaseWeatherIcon from "@/components/BaseComponents/BaseWeatherIcon";
```
Comment (Virtuosofriend) [inline src/components/Stations/components/StationTableColumns.tsx:1]: Excellent call to move all this to a different file!

File: src/hooks/useAppStore.ts [hunk 12]
```diff
@@ -1,13 +1,24 @@
+import {
+    getFavouritesStationList,
+    getShowFavouriteStations,
+    storeFavouriteStationsList,
+    storeShowFavouriteStations
+} from "@/utils/localStorage";
 import { create } from "zustand";
 
 type AppStore = {
```
Comment (Virtuosofriend) [inline src/hooks/useAppStore.ts:9]: Do you feel that this type should be added in the types file or leave it here?

File: src/components/Stations/components/StationTableColumns.tsx (hunk 9 above)
Comment (dimitrantz) [reply src/components/Stations/components/StationTableColumns.tsx]: yep makes sense, I've updated it

**Review by Virtuosofriend — APPROVED** [review]
